crates/loans/src/lib.rs

```diff
@@ -1033,42 +1029,14 @@ pub mod pallet {
             let who = ensure_signed(origin.clone())?;
             // This function is an almost 1:1 duplicate of the logic in `do_redeem`.
             // It could be refactored to compute the redeemable underlying
-            // with `Self::recompute_underlying_amount(&Self::free_lend_tokens(asset_id, &who)?)?`
-            // but that would cause the `accrue_interest_works_after_redeem_all` unit test to fail with
-            // left: `1000000000003607`,
-            // right: `1000000000003608`'
-
-            // Chaining `calc_underlying_amount` and `calc_collateral_amount` continuously decreases
-            // an amount because of rounding down, and having the current function call `do_redeem`
-            // would perform three conversions: lend_token -> token -> lend_token -> token.
-            // Calling `do_redeem_voucher` directly only performs one conversion: lend_token -> token,
-            // avoiding this edge case.
-            // TODO: investigate whether it is possible to implement the conversion functions
-            // with guarantees that this always holds:
-            // `calc_underlying_amount(calc_collateral_amount(x)) = calc_collateral_amount(calc_underlying_amount(x))`
-            // Use the `converting_to_and_from_collateral_should_not_change_results` unit test to achieve this.
-            // If there are leftover lend_tokens after a `redeem_all` (because of rounding down), it would make it
-            // impossible to enforce "collateral toggle" state transitions.
-            Self::ensure_active_market(asset_id)?;
+            let free = Self::free_lend_tokens(asset_id, &who)?;
+            let reserved = Self::reserved_lend_tokens(asset_id, &who)?;
+            let used_collateral = Amount::max(&free, &reserved)?;
```

Comment:
An account can't have both `free` and `reserved` lend tokens so this is correct but is not very easy to understand at first glance. Would be good to add a comment.

Reply:
@sander2 Actually the caller could have both `free` and `reserved` lend tokens if they just mint and then lock some of their balance as vault collateral. If they lock more vault collateral than the free collateral, this call will end up failing although it should just redeem the free tokens.

Reply:
> @sander2 Actually the caller could have both `free` and `reserved` lend tokens if they just mint and then lock some of their balance as vault collateral. If they lock more vault collateral than the free collateral, this call will end up failing although it should just redeem the free tokens.

Oh shit I was assuming that free and reserved would be mutually exclusive.. That's not hard to fix but that means I reviewed some of the code with incorrect assumptions :(